[CI] CoordinatorTests.testDoesNotPerformElectionWhenRestartingFollower failure

**Build scan**:
https://gradle-enterprise.elastic.co/s/vhneniggttxdq

**Repro line**:

```
./gradlew ':server:test' --tests "org.elasticsearch.cluster.coordination.CoordinatorTests.testDoesNotPerformElectionWhenRestartingFollower" -Dtests.seed=D5EE617ADD52F575 -Dtests.security.manager=true -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=de-CH -Dtests.timezone=Asia/Manila -Druntime.java=8
```

**Reproduces locally?**:
```
./gradlew ':server:test' --tests "org.elasticsearch.cluster.coordination.CoordinatorTests.testDoesNotPerformElectionWhenRestartingFollower" -Dtests.seed=D5EE617ADD52F575 -Dtests.security.manager=true -Dtests.locale=de-CH -Dtests.timezone=Asia/Manila -Druntime.java=8
```
reproduces locally for me.

**Applicable branches**:

Failed in 7.x only (but only failed once).

**Failure history**:

https://build-stats.elastic.co/goto/0d9728f81d0a7966d908aae65e4751d4
Failed once only recently.

**Failure excerpt**:

```
org.elasticsearch.cluster.coordination.CoordinatorTests > testDoesNotPerformElectionWhenRestartingFollower FAILED
    java.lang.AssertionError: node1 is a follower of node0
    Expected: is <FOLLOWER>
         but: was <CANDIDATE>
        at __randomizedtesting.SeedInfo.seed([D5EE617ADD52F575:489C4247AC88A7A0]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.cluster.coordination.AbstractCoordinatorTestCase$Cluster.stabilise(AbstractCoordinatorTestCase.java:542)
        at org.elasticsearch.cluster.coordination.AbstractCoordinatorTestCase$Cluster.stabilise(AbstractCoordinatorTestCase.java:502)
        at org.elasticsearch.cluster.coordination.CoordinatorTests.testDoesNotPerformElectionWhenRestartingFollower(CoordinatorTests.java:1561)
```

--- a/test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java
@@ -381,6 +381,14 @@ public class AbstractCoordinatorTestCase extends ESTestCase {
                     }
                 }
 
+                // Drain all the disruption events during cool down period until all nodes are healthy.
+                // This prevents some edge cases where a disruption is scheduled to run some time after the stabilization period
+                // (i.e. black-holed connections throw a disconnected exception after 1 day), preventing the cluster to reach
+                // a stable state. See #61711 for a particular instance of this scenario.
+                if (finishTime != -1 && (disconnectedNodes.isEmpty() == false || blackholedNodes.isEmpty() == false)) {
+                    finishTime = deterministicTaskQueue.getLatestDeferredExecutionTime();
+                }
+
                 try {
                     if (finishTime == -1 && randomBoolean() && randomBoolean() && randomBoolean()) {
                         final ClusterNode clusterNode = getAnyNodePreferringLeaders();
@@ -445,12 +453,14 @@ public class AbstractCoordinatorTestCase extends ESTestCase {
                                 }
                                 break;
                             case 1:
-                                if (clusterNode.disconnect()) {
+                                // Avoid disruptions during cool down period
+                                if (finishTime == -1 && clusterNode.disconnect()) {
                                     logger.debug("----> [runRandomly {}] disconnecting {}", step, clusterNode.getId());
                                 }
                                 break;
                             case 2:
-                                if (clusterNode.blackhole()) {
+                                // Avoid disruptions during cool down period
+                                if (finishTime == -1 && clusterNode.blackhole()) {
                                     logger.debug("----> [runRandomly {}] blackholing {}", step, clusterNode.getId());
                                 }
                                 break;

--- a/test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java
@@ -371,21 +371,19 @@ public class AbstractCoordinatorTestCase extends ESTestCase {
                 final int thisStep = step; // for lambdas
 
                 if (randomSteps <= step && finishTime == -1) {
-                    finishTime = deterministicTaskQueue.getLatestDeferredExecutionTime();
                     if (coolDown) {
+                        // Heal all nodes BEFORE finishTime is set so it can take into account any pending disruption that
+                        // would prevent the cluster to reach a stable state after cooling down. Additionally, avoid any new disruptions
+                        // to happen in this phase.
+                        // See #61711 for a particular instance where having unhealthy nodes while cooling down can be problematic.
+                        disconnectedNodes.clear();
+                        blackholedNodes.clear();
                         deterministicTaskQueue.setExecutionDelayVariabilityMillis(DEFAULT_DELAY_VARIABILITY);
                         logger.debug("----> [runRandomly {}] reducing delay variability and running until [{}ms]", step, finishTime);
                     } else {
                         logger.debug("----> [runRandomly {}] running until [{}ms] with delay variability of [{}ms]", step, finishTime,
                             deterministicTaskQueue.getExecutionDelayVariabilityMillis());
                     }
-                }
-
-                // Drain all the disruption events during cool down period until all nodes are healthy.
-                // This prevents some edge cases where a disruption is scheduled to run some time after the stabilization period
-                // (i.e. black-holed connections throw a disconnected exception after 1 day), preventing the cluster to reach
-                // a stable state. See #61711 for a particular instance of this scenario.
-                if (finishTime != -1 && (disconnectedNodes.isEmpty() == false || blackholedNodes.isEmpty() == false)) {
                     finishTime = deterministicTaskQueue.getLatestDeferredExecutionTime();
                 }
 

